mineral_deposit_tracking/src/mineral_deposit_tracker.cpp

```diff
@@ -0,0 +1,124 @@
+#include <stsl_interfaces/msg/mineral_deposit_array.hpp>
+#include <stsl_interfaces/srv/reset_mineral_deposit_tracking.hpp>
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include "kalman_filter.hpp"
+
+namespace mineral_deposit_tracking
+{
+
+class MineralDepositTracker : public rclcpp::Node
+{
+public:
+  explicit MineralDepositTracker(const rclcpp::NodeOptions & options)
+  : rclcpp::Node("mineral_deposit_tracker", options),
+    tf_buffer_(get_clock()),
+    tf_listener_(tf_buffer_),
+    filter_(Eigen::Matrix2d::Identity(),
+      Eigen::Matrix2d::Identity() * 1e-4, Eigen::Matrix2d::Identity())
+  {
+    tracked_deposit_publisher_ = create_publisher<geometry_msgs::msg::PoseStamped>(
+      "~/tracked_deposit", rclcpp::SystemDefaultsQoS());
+    deposit_subscription_ = create_subscription<stsl_interfaces::msg::MineralDepositArray>(
+      "~/sensed_deposits", rclcpp::SystemDefaultsQoS(),
+      std::bind(&MineralDepositTracker::DepositMeasurementCallback, this, std::placeholders::_1));
+    reset_service_ = create_service<stsl_interfaces::srv::ResetMineralDepositTracking>(
+      "~/reset", std::bind(
+        &MineralDepositTracker::ResetCallback, this, std::placeholders::_1,
+        std::
+        placeholders::_2));
+  }
+
+private:
+  tf2_ros::Buffer tf_buffer_;
+  tf2_ros::TransformListener tf_listener_;
+  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr tracked_deposit_publisher_;
+  rclcpp::Subscription<stsl_interfaces::msg::MineralDepositArray>::SharedPtr deposit_subscription_;
+  rclcpp::Service<stsl_interfaces::srv::ResetMineralDepositTracking>::SharedPtr reset_service_;
+  int deposit_id_;
+  KalmanFilter filter_;
+
+  void DepositMeasurementCallback(const stsl_interfaces::msg::MineralDepositArray::SharedPtr msg)
+  {
+    try {
+      filter_.TimeUpdate();
+      const auto found_deposit = std::find_if(
+        msg->deposits.begin(), msg->deposits.end(), [this](const auto & deposit) {
+          return deposit.id == deposit_id_;
+        });
+      if (found_deposit == msg->deposits.end()) {
+        return;
+      }
+      const auto & detection = *found_deposit;
+
+      const Eigen::Vector3d sensor_frame_position = detection.range * Eigen::Vector3d{std::cos(
+          detection.heading), std::sin(detection.heading), 0.0};
+
+      const auto transform =
+        tf_buffer_.lookupTransform(
+        "map", msg->header.frame_id, msg->header.stamp, rclcpp::Duration::from_seconds(
+          0.1));
+
+      Eigen::Vector3d map_frame_position;
+
+      tf2::doTransform(sensor_frame_position, map_frame_position, transform);
+
+      const Eigen::Vector2d measurement{map_frame_position.x(), map_frame_position.y()};
+
+      const Eigen::Matrix2d covariance = Eigen::Matrix2d::Identity() * 0.01;
+
+      filter_.MeasurementUpdate(measurement, covariance);
+
+      const auto estimated_position = filter_.GetEstimate();
+
+      geometry_msgs::msg::PoseStamped output_msg;
+      output_msg.header.stamp = now();
```

Comment:
Should this be the timestamp of the message?

Reply:
I don't know if it matters much. I don't see a strong connection between a specific message and the published estimate, so I set to `now()`.